### PR TITLE
rgw: Break Full and Incremental sync out into their own classes

### PR DESCRIPTION
**Pull request by adamemerson**
That way we don't end up going down the wrong fork when Full is finishing up.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_data_sync.cc [hunk 1]
```diff
@@ -1774,20 +1806,70 @@ class RGWDataSyncShardCR : public RGWCoroutine {
...
+public:
+  RGWDataSyncShardCR(RGWDataSyncCtx* const _sc, const rgw_pool& pool,
+                     const uint32_t shard_id,
+		     rgw_data_sync_marker& sync_marker,
+                     RGWSyncTraceNodeRef& tn,
+		     bool* /* XXX We were passing this in as reset_backoff but we
+			      never did anything with it. If we should be doing
+			      something with it, we should put that in. If not
+			      we should just remove it from the supplied
+			      parameters. */)
```
Comment (cbodley) [inline src/rgw/rgw_data_sync.cc:1830]: DataSyncShardCR is wrapped in a BackoffControlCR, which will keep retrying DataSyncShardCR after it returns an error

if we have two rgws running in the same zone, we'd expect one of them to grab the lock on this shard and process it, while the other would keep looping on `EBUSY`

so the backoff behavior is mostly controlling how we poll this lock - we want to keep backing off until we succeed in acquiring it

if we're moving the management of `RGWContinuousLeaseCR` into `RGWDataSyncShardCR` as i suggested above, that should make it easier to manage this flag. but you're also welcome to leave this as a TODO comment

**Review by cbodley — COMMENTED** [review]
👍

File: src/rgw/rgw_data_sync.cc [hunk 2]
```diff
@@ -1487,77 +1467,43 @@ class RGWDataSyncShardCR : public RGWCoroutine {
...
+  RGWDataBaseSyncShardCR(RGWDataSyncCtx *const _sc,
+			 const rgw_pool& pool,
+			 const uint32_t shard_id,
+			 rgw_data_sync_marker& sync_marker,
+			 const RGWSyncTraceNodeRef& tn)
+    : RGWCoroutine(_sc->cct), sc(_sc), pool(pool), shard_id(shard_id),
+      sync_marker(sync_marker), tn(tn)
+    {}
+
+  ~RGWDataBaseSyncShardCR() override {
```
[the remark this answers is not in the document]
Comment (adamemerson) [reply src/rgw/rgw_data_sync.cc:1480]: Yeah. This is what I meant when I said I got confused by the 'returns' in the original. I had thought that the Full path returned then the coroutine was torn down then spawned again, so I was under the impression things shared code but not data.

I'm refactoring the cache, lease, and a couple other things back into the original.

[the remark this answers is not in the document]
Comment (adamemerson) [reply src/rgw/rgw_data_sync.cc]: Done.

File: src/rgw/rgw_data_sync.cc (hunk 1 above)
Comment (adamemerson) [reply src/rgw/rgw_data_sync.cc]: Stuck a TODO on the flag.

**Comment by adamemerson** [comment]
@cbodley Is there an upstream tracker for this or should I open one?

**Comment by cbodley** [comment]
> @cbodley Is there an upstream tracker for this or should I open one?

@adamemerson if you think this is worth backporting to quincy/pacific, sure

**Review by cbodley — APPROVED** [review]
fixed some bugs and cleaned up the locking. it's passing multisite tests

**Comment by cbodley** [comment]
@adamemerson could you please look over the commits i added?

**Review by adamemerson — COMMENTED** [review]
@cbodley Looks good to me.

**Comment by mattbenjamin** [comment]
> fixed some bugs and cleaned up the locking. it's passing multisite tests

superb!

**Comment by cbodley** [comment]
mostly passed qa in https://pulpito.ceph.com/cbodley-2022-08-11_19:52:36-rgw-wip-coroutine-cleaver-distro-default-smithi/, but multisite tests all failed due to https://github.com/ceph/ceph/pull/47411

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
verified manually with multisite tests, only failure was from `test_multi.test_version_suspended_incremental_sync` tracked in https://tracker.ceph.com/issues/55179 👍

**Pull request merged**